File: src/main/java/com/dude/dms/ui/views/DocsView.java
package com.dude.dms.ui.views;

import com.dude.dms.backend.brain.parsing.PdfToDocParser;
import com.dude.dms.backend.data.docs.Doc;
import com.dude.dms.backend.data.history.DocHistory;
import com.dude.dms.backend.data.Tag;
import com.dude.dms.backend.service.DocService;
import com.dude.dms.backend.service.TagService;
import com.dude.dms.ui.Const;
import com.dude.dms.ui.MainView;
import com.dude.dms.ui.builder.BuilderFactory;
import com.dude.dms.ui.components.tags.TagContainer;
import com.dude.dms.ui.converters.LocalDateConverter;
import com.helger.commons.io.file.FileHelper;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.html.Anchor;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.router.*;
import com.vaadin.flow.server.StreamResource;
import org.springframework.beans.factory.annotation.Autowired;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;

import static com.dude.dms.backend.brain.OptionKey.DOC_SAVE_PATH;

@Route(value = Const.PAGE_DOCS, layout = MainView.class)
@RouteAlias(value = Const.PAGE_ROOT, layout = MainView.class)
@PageTitle("Docs")
public class DocsView extends HistoricalCrudView<Doc, DocHistory> implements HasUrlParameter<String> {

    private final BuilderFactory builderFactory;

    private final DocService docService;
    private final TagService tagService;

    private String param;

    @Autowired
    public DocsView(BuilderFactory builderFactory, DocService docService, TagService tagService, PdfToDocParser pdfToDocParser) {
        this.builderFactory = builderFactory;
        this.docService = docService;
        this.tagService = tagService;
        UI ui = UI.getCurrent();
        pdfToDocParser.addEventListener("docs", success -> {
            if (success) {
                ui.access(this::fillGrid);
            }
        });
        addColumn("Date", doc -> LocalDateConverter.convert(doc.getDocumentDate()));
        addComponentColumn("Tags", doc -> new TagContainer(doc.getTags()));
        addComponentColumn("", this::createGridActions);
        addColumn("GUID", Doc::getGuid);

        grid.addItemDoubleClickListener(event -> builderFactory.docs().imageDialog(event.getItem()).build().open());
    }

    private HorizontalLayout createGridActions(Doc doc) {
        Path path = Paths.get(DOC_SAVE_PATH.getString(), "pdf", doc.getGuid() + ".pdf").toAbsolutePath();
        File file = path.toFile();
        Anchor download = new Anchor();
        download.add(new Button(VaadinIcon.FILE_TEXT.create()));
        download.setEnabled(false);
        if (file.exists()) {
            download.setEnabled(true);
            download.setHref(new StreamResource("pdf.pdf", () -> FileHelper.getInputStream(file)));
            download.getElement().setAttribute("download", true);
        }

        return new HorizontalLayout(
                new Button(VaadinIcon.TEXT_LABEL.create(), e -> builderFactory.docs().textDialog(doc).build().open()),
                download,
                new Button(VaadinIcon.EDIT.create(), e -> builderFactory.docs().editDialog(doc).withEventListener(() -> grid.getDataProvider().refreshAll()).build().open())
        );
    }

    private void fillGrid() {
        if (param != null && !param.isEmpty()) {
            String[] parts = param.split(":");
            if ("tag".equalsIgnoreCase(parts[0])) {
                Optional<Tag> tag = tagService.findByName(parts[1]);
                tag.ifPresent(t -> grid.setItems(docService.findByTag(t)));
            }
        } else {
            grid.setItems(docService.findAll());
        }
    }

    @Override
    public void setParameter(BeforeEvent beforeEvent, @OptionalParameter String t) {
        param = t;
        fillGrid();
    }
}

File: src/main/kotlin/com/dude/dms/ui/views/DocsView.kt
package com.dude.dms.ui.views

import com.dude.dms.backend.containers.DocContainer
import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.data.docs.Attribute
import com.dude.dms.backend.data.docs.Doc
import com.dude.dms.backend.service.DocService
import com.dude.dms.brain.events.EventManager
import com.dude.dms.brain.events.EventType
import com.dude.dms.brain.options.Options
import com.dude.dms.brain.t
import com.dude.dms.ui.Const
import com.dude.dms.ui.components.cards.DocCard
import com.dude.dms.ui.components.dialogs.DocImageDialog
import com.dude.dms.ui.components.misc.DocSearchBar
import com.dude.dms.ui.components.misc.ViewPageSelector
import com.dude.dms.utils.docCard
import com.dude.dms.utils.queryService
import com.dude.dms.utils.searchBar
import com.dude.dms.utils.viewPageSelector
import com.github.mvysny.karibudsl.v10.*
import com.vaadin.flow.component.Text
import com.vaadin.flow.component.UI
import com.vaadin.flow.component.html.Div
import com.vaadin.flow.component.icon.VaadinIcon
import com.vaadin.flow.component.orderedlayout.FlexComponent
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.router.*
import org.springframework.data.domain.PageRequest
import kotlin.streams.toList


@Route(value = Const.PAGE_DOCS, layout = MainView::class)
@RouteAlias(value = Const.PAGE_ROOT, layout = MainView::class)
@PageTitle("Docs")
class DocsView(
    private val docService: DocService,
    private val eventManager: EventManager
) : VerticalLayout(), HasUrlParameter<String?>, BeforeLeaveObserver {

    private val viewUI = UI.getCurrent()

    private val itemContainer: Div

    private val docSearchBar: DocSearchBar

    private lateinit var pageSelector: ViewPageSelector

    private lateinit var itemCount: Text

    private var filter = ""

    init {
        eventManager.register(this, Doc::class, EventType.CREATE) { fill(viewUI) }
        eventManager.register(this, Doc::class, EventType.UPDATE) { updateDoc(it, viewUI) }
        eventManager.register(this, Doc::class, EventType.DELETE) { deleteDoc(it, viewUI) }
        eventManager.register(this, Tag::class, EventType.CREATE, EventType.UPDATE, EventType.DELETE) { fill(viewUI) }
        eventManager.register(this, Attribute::class, EventType.CREATE, EventType.UPDATE, EventType.DELETE) { fill(viewUI) }

        style["paddingTop"] = "0px"
        style["paddingBottom"] = "0px"
        setHeightFull()

        docSearchBar = searchBar {
            style["position"] = "sticky"
            style["top"] = "0"
            style["zIndex"] = "1"
            style["borderBottom"] = "1px solid var(--lumo-contrast-10pct)"
            style["backgroundColor"] = "var(--lumo-base-color)"
            onChange = {
                this@DocsView.filter = it
                fill()
            }
        }
        itemContainer = div {
            setWidthFull()
            style["display"] = "flex"
            style["flexWrap"] = "wrap"
            style["paddingBottom"] = "40px"
        }
        horizontalLayout {
            setWidthFull()
            justifyContentMode = FlexComponent.JustifyContentMode.CENTER
            alignItems = FlexComponent.Alignment.CENTER
            style["position"] = "sticky"
            style["bottom"] = "0"
            style["marginTop"] = "auto"
            style["borderTop"] = "1px solid var(--lumo-contrast-10pct)"
            style["backgroundColor"] = "var(--lumo-base-color)"

            label("${t("zoom")}:")
            iconButton(VaadinIcon.MINUS_CIRCLE.create()) {
                onLeftClick { shrink() }
            }
            iconButton(VaadinIcon.PLUS_CIRCLE.create()) {
                onLeftClick { grow() }
            }
            div { width = "2em" }
            pageSelector = viewPageSelector()
            div { width = "2em" }
            itemCount = text(t("items") + ":")
        }

        pageSelector.setChangeListener { fill() }
        fill()
    }

    private fun updateDoc(doc: Doc, ui: UI) {
        ui.access {
            itemContainer.children.toList().filterIsInstance<DocCard>().firstOrNull { it.docContainer.doc?.guid == doc.guid }?.fill()
        }
    }

    private fun deleteDoc(doc: Doc, ui: UI) {
        ui.access {
            itemContainer.children.toList().filterIsInstance<DocCard>().firstOrNull { it.docContainer.doc?.guid == doc.guid }?.let {
                itemContainer.remove(it)
            }
        }
    }

    private fun grow() {
        val options = Options.get()
        if (options.view.docCardSize < 400) {
            options.view.docCardSize += 10
            options.save()
            itemContainer.children.filter { it is DocCard }.forEach { (it as DocCard).resize() }
        }
    }

    private fun shrink() {
        val options = Options.get()
        if (options.view.docCardSize > 100) {
            options.view.docCardSize -= 10
            options.save()
            itemContainer.children.filter { it is DocCard }.forEach { (it as DocCard).resize() }
        }
    }

    private fun fill() {
        val docs = docService.findByFilter(filter, PageRequest.of(pageSelector.page, pageSelector.pageSize.value))
        itemContainer.removeAll()
        pageSelector.items = docService.countByFilter(filter).toInt()
        itemCount.text = "${t("items")}: ${docs.size}"
        docs.forEach { itemContainer.docCard(DocContainer(it)) { addClickListener { DocImageDialog(docContainer).open() } } }
    }

    private fun fill(ui: UI) {
        ui.access { fill() }
    }

    override fun setParameter(beforeEvent: BeforeEvent, @OptionalParameter t: String?) {
        if (!t.isNullOrEmpty()) {
            val parts = t.split(":").toTypedArray()
            when {
                "tag".equals(parts[0], ignoreCase = true) -> docSearchBar.filter.value = "${t("tag")} = ${parts[1]}"
                "query".equals(parts[0], ignoreCase = true) -> docSearchBar.filter.value = queryService.load(parts[1].toLong())?.filter ?: ""
                else -> docSearchBar.filter.clear()
            }
        }
    }

    override fun beforeLeave(event: BeforeLeaveEvent?) {
        eventManager.unregister(this)
    }
}
